### PR TITLE
Introduce traits

### PR DESCRIPTION
**Pull request by uncleDecart**

### CODE REVIEW

**Comment by uncleDecart** [comment]
This solves #9

**Review by deitch — COMMENTED** [review]
Reading this through, the irony is that the file that defines the traits - `src/traits.rs` - is the last one lexicographically. So I was going through this, trying to understand it, the whole time thinking, "this would be easier if I knew the definition of traits, but where is it?" 😆 

My instinct usually is to put interface definitions (I know, "interface" is a go or java word, not rust, but it describes the concept much better) in their own place, i.e. their own directory. That may not be rust-ish (I don't know). But I like isolating those into an easy-to-find and -read space. So take it is a light suggestion, rather than a strong one.

Next is what the traits are. I see three main ones: `Value`, `PersistValue`, `Notifier`. Where do these fill in? Are these for a driver for the server, i.e. storage? Unlikely, since it has `Notifier`. So maybe `PersistValue` is for backend storage? Or `Notifier` too? So what part of the architecture do these slot into, or what prior concrete class (yeah, "struct", I know) do these replace, and those become actual implementations of?

Finally, documentation. We should document what it looks like under the covers, and how to either replace parts or slide in alternative implementations.

Tying the last two together, I would suggest not to bother answering the above questions directly in the comments - unless they are unclear - but rather in the docs themselves.

**Comment by deitch** [comment]
Essentially, work under the assumption that the next person to work on this is not you and has no time to be trained by you, but has to replace _some part_ of it. How can we jumpstart them?

**Comment by uncleDecart** [comment]
@deitch check out this [doc](https://github.com/uncleDecart/nkv/blob/d57992e94a24752bf5e5d2aea2b83d0f0da15848/docs/DESIGN.md)

File: docs/DESIGN.md [hunk 1]
```diff
@@ -20,14 +20,24 @@ handles the channel between server and client to notify latter if anybody change
...
-You can call those components interfaces, and implement them differently to tailor to your particular 
-case, i.e. use mysql for persist value or unix domain socket for `Notfifier` and TCP for `Server` 
 
 From the flow diagram you can see how `NotifyKeyValue` processes requests.
 
 ![nkv flow diagram](../imgs/nkv_flow.drawio.png)
 
+# How can I tailor `nkv` to my specific needs?
+
+Default implementation of `nkv` might not be suitable to your particular use case, f.e. you want to store
```
Comment (deitch) [inline docs/DESIGN.md:32]: `f.e.`? I think you mean as in "for example"?

File: docs/DESIGN.md [hunk 2]
```diff
@@ -20,14 +20,24 @@ handles the channel between server and client to notify latter if anybody change
...
+values in a mysql database or you do not want to keep them at all, or you want to notify your clients via 
+unix domain socket or zmq socket because your existing system relies on it or you think it's better, etc.
+
+Well, you can fairly easily do that. We abstract `PersistValue` and `Notifier` as traits. You can see it in
+`src/traits.rs` file. `NotifyKeyValue` structure is using generics to abstract from concrete `PersistValue` and
```
Comment (deitch) [inline docs/DESIGN.md:37]: link to the file?

File: docs/DESIGN.md (hunk 1 above)
Comment (uncleDecart) [reply docs/DESIGN.md:32]: yep, can expand it :)

File: docs/DESIGN.md (hunk 2 above)
Comment (uncleDecart) [reply docs/DESIGN.md:37]: ah yes, it'll be better :D

Comment (uncleDecart) [reply docs/DESIGN.md]: fixed

File: docs/DESIGN.md (hunk 1 above)
Comment (uncleDecart) [reply docs/DESIGN.md:32]: fixed

**Comment by deitch** [comment]
I am focusing primarily on the design doc. This was harder to do inline.

If I got the doc, you are saying that: nkv is a key-value store with notification. It primarily is a server, which exposes a simple API over a socket (network or file/unix domain), as well as a convenient client. Client isn't necessary, you can use the API directly via curl or anything else, if you want, but client - sdk and a wrapped CLI - makes it a lot more convenient.

So far so good?

The client just uses the sdk to wrap the API, so nothing much to discuss there.

Ah, yes, we should have an explicit document for the API, either hand-written or autogenerated.

The server "bridges" between the API over the socket and the storage engine. Now we are getting deeper into what is in the `design.md` doc.

I think there is one primary trait, `NotifyKeyValue`. This could just as easily have been named `NkvStorage`. It implements the storage engine? That is what I first thought, but `NotifyKeyValue` is not a trait, but a concrete struct. So `NotifyKeyValue` is an in-memory  structure that contains a map, where the keys are strings, and the values are `Value struct`, where the properties of `Value` are a `PersistValue` (trait) and a `Notifier` (trait).

So each value in the map is an individual storage engine (i.e. `PersistValue`), and an implementation that can notify subscribers about updates for that particular string key.

* First, is that it? It is somewhat hard to understand the structure from the doc. It reads a bit like going from your understanding of how it works to how it is built, rather than the other way around.
* Second, is this actually a "persist" value? Isn't that a choice? We might use a storage engine that writes everything to a database, another that does an mmapped file, another that saves it in Redis, another that is entirely in memory?
* Third, why do it at the individual item level? Why not abstract the whole storage engine as a trait? So one engine would be backed by Redis, another by memory, another by mmapped files, another by custom files? Sort of the way that eden does it?

**Comment by uncleDecart** [comment]
> I think there is one primary trait, NotifyKeyValue. This could just as easily have been named NkvStorage. It implements the storage engine? That is what I first thought, but NotifyKeyValue is not a trait, but a concrete struct. So NotifyKeyValue is an in-memory structure that contains a map, where the keys are strings, and the values are Value struct, where the properties of Value are a PersistValue (trait) and a Notifier (trait).

I think if you rename PersistValue to StoragePolicy it'll become clearer. So in a nutshell NkvStorage (I like that name too, think we should keep it) does nothing but caching Generic Value. Value is another generic structure which consist of StoragePolicy and Notifier, those are 2 generic objects implementing traits. Meaning I already have compile time abstraction of Notification policies and Storage policies (maybe we should rename notifier and NotificationPolicy) . Plainly when creating NkvStorage I have to specify how I will notify my clients and how will I store the values. If we want to abstract Caching policies we can create a trait for that (i.e. store in cache only used values and offload others), then we will have LRUCachedTrie and NormalTrie, etc. etc. etc. As of now introducing this traits we have what you're talking about in third point, it'll look like this:

```rust
        let mut fs_tcp_nkv = NotifyKeyValue::<FileStorage, TcpNotifier>::new(path)?;
        let mut redis_tcp_nkv = NotifyKeyValue::<RedisStorage, TcpNotifier>::new(path)?;
        let mut mysql_uds_nkv = NotifyKeyValue::<MySqlStorage, UnixSockNotifier>::new(path)?;
```

**Comment by deitch** [comment]
Continuing the conversation from earlier today. The brief notes I recall.

* `Server` = the primary engine that runs. It listens on a socket - Unix domain or TCP - for requests that conform to the API. It takes those requests (maybe authentication, etc.), validates them, and then sends the request over a channel to the nkvcore (or whatever you want to call it). Core receives on channel, and handles put/delete/get by calling whichever concrete implementation of `StorageEngine` is connected to it.
* `Client` = just talks over socket to server. Convenient wrapper around REST API

Within the server, there are a few pieces. Naming is just a suggestion.

* frontend - part of the server that handles socket communications, and translates between socket and channels. Also handles authentication (if we ever implement it) and API validation, etc. Also responsible for notifying nkvcore over channel about subscribers that went away. E.g. had a subscriber on a socket, it got shut down. frontend should tell core, "unsubscribe X".
* `NkvCore` - primary library surface for server functionality. Does not listen on sockets, but rather on rust channels. Entirely expected to be in-process. Handles:
  * cache engine - if cache is selected, then this handles cache management, interfacing with the cache before/after the storage engine
  * storage engine - could be called `StorageEngine` or `StorageBackend` or even just `Storage`. Whatever you call it, a trait, that any concrete backend should implement. Should be able to handle put(k,v)/delete(k)/get(k)
  * notifier engine - keeps track of all of the subscribers to particular keys or patterns. After each update or delete, goes through the subscribers, creates a list of those to notify, and sends the list on the channel to the frontend, which handles how the actual notification happens.

From core's perspective, it doesn't know anything about notifications, about sockets, about APIs. It has a simple set of channels for receiving requests/sending responses and for sending notification messages. It is possible that the response and notification channels are the same, since there is not much difference between the response to "GET x" and "tell me when x changes".

**Comment by uncleDecart** [comment]
Pheewww, So I rewrote docs and traits to be aligned with the conversation we had before. Didn't introduce caching trait yet, still caches everything in memory. Let me know if new version of doc is clearer or if I missed something.

**Comment by deitch** [comment]
This is really nice, and easy to understand. I didn't look at how the subscriptions and notifications are handled inside core, because you always can improve that. It is fully internal. I just have some suggestions on the doc language. I will comment inline.

File: docs/DESIGN.md [hunk 3]
```diff
@@ -1,33 +1,37 @@
...
-From the flow diagram you can see how `NotifyKeyValue` processes requests.
+Think about `nkv` as a black box implementing following standard key-value storage API with a twist:
 
-![nkv flow diagram](../imgs/nkv_flow.drawio.png)
+- Get value
+- Put value (update in case value existed)
+- Delete value
+- Subscribe to value updates
+- Unsubscribe to value updates
+
```
Comment (deitch) [inline docs/DESIGN.md:10]: How about something like this, a slight variant on your language?

nkv is 3 components:

* core library that provides key-value storage and subscription management, implementing all of the above core functionality, communicated via rust channels, usable by rust programs
* standalone server that wraps the core library to provide a socket-based API accessible to any client that communicates with the API over the socket
* standalone client that simplifies communicating with the server over the socket

The `NkvCore` library allows you to put, get, delete values and subscribe to recieve notifications on
value updates via Rust channels. When instantiated, it receives a parameter called `StorageEngine`. This is what stores the actual data. `StorageEngine` is a trait, defined [here](../src/traits.rs), and can be implemented via drivers to different storage engines, for example, memory, persistent on disk, mysql database, or anything else that can implement get, put and delete.

Comment (uncleDecart) [reply docs/DESIGN.md:10]: I like it, do you want to add it after 5 endpoints of instead?

Comment (uncleDecart) [reply docs/DESIGN.md:8]: Rewrote the stuff, I think it's clearer now

**Comment by uncleDecart** [comment]
@deitch are you okay with merging this?

**Comment by deitch** [comment]
Yup

**Pull request merged**